# Microexon Code paper


This repository contains the scripts and model predictions used in the analysis of autism-associated genetic variation and its impact on brain microexons, as detailed in our paper "An expanded repertoire of brain microexons is directly impacted by autism-associated genetic variation"

## Description

Our study presents an integrative analysis of an expanded repertoire of neuronal differential microexons and the development of a machine learning model that delineates sequence elements critical for regulating microexon splicing. We reveal a landscape of predominantly rare human genetic variation that convergently impacts a subset of microexons in genes with known and unknown links to nervous system biology and neurodevelopmental disorders.

## Contents

- `microexon-code`: This directory contains the computational workflow to perform predictions.
- `DataAnalysis/`: This directory contains all the scripts used for data analysis.
- `OutputTables/`: This directory contains the model predictions for all single nucleotide variants (SNVs) and autism spectrum disorder (ASD) indels identified in our study.

## Output Variant predictions

The predictions for all variants evaluated by our model as part of the analysis carried out in the paper are can be found at `OutputTables/ASD_indels_and_ISM_SNVs.deta_code_preditions.tsv`. Detail descriptions for this table and other complementary tables are found at `OutputTables/README.md`.


## License

This project is licensed under the MIT License - see the LICENSE.md file for details.

## Citation

If you use the scripts or data provided in this repository, please cite our paper:
Parada, G.E., Bretschneider, H., et al. (2024). An expanded repertoire of brain microexons is directly impacted by autism-associated genetic variation. *Journal Name*, *Volume*(Issue), pages.

## Contact

For any questions or support, please open an issue in the repository or contact the corresponding authors at their email addresses provided in the paper.
